lib/concurrent-merkle-tree/src/merkle_roll.rs

```diff
@@ -196,6 +196,91 @@ impl<const MAX_DEPTH: usize, const MAX_BUFFER_SIZE: usize> MerkleRoll<MAX_DEPTH,
         Ok(node)
     }
 
+    /// Append subtree to current tree
+    pub fn append_subtree(
+        &mut self,
+        subtree_root: Node,
+        subtree_rightmost_leaf: Node,
+        subtree_rightmost_index: u32,
+        subtree_rightmost_proof: Vec<Node>,
+    ) -> Result<Node, CMTError> {
+        check_bounds(MAX_DEPTH, MAX_BUFFER_SIZE);
+        // TODO: consider adding check that the subtree is not empty (but will omit for now)
+        //       note: it will be more time consuming to check that the subtree_root is not a root to an empty tree, because it requires applying O(depth) hashes of trivial nodes
+        if self.rightmost_proof.index >= 1 << MAX_DEPTH {
+            return Err(CMTError::TreeFull);
+        }
+
+        if self.rightmost_proof.index == 0 {
+            // TODO(sorend): Consider if we want to be able to initialize a tree with a subtree append (I think it would be cool, but is a little complex). write an equivallent function for subtree_append: initialize_tree_from_subtree_append
+            // For now
+            return Err(CMTError::CannotInitializeWithSubtreeAppend);
+        }
+
+        // Confirm that subtree_rightmost_proof is valid
+        if recompute(
+            subtree_rightmost_leaf,
+            &subtree_rightmost_proof[..],
+            subtree_rightmost_index - 1,
+        ) != subtree_root
+        {
+            return Err(CMTError::InvalidProof);
+        }
+
+        let leaf = subtree_rightmost_leaf.clone();
+        let intersection = self.rightmost_proof.index.trailing_zeros() as usize;
+
+        // @dev: At any given time (other than initialization), there is only one valid size of subtree that can be appended
+        if subtree_rightmost_proof.len() != intersection {
```

Comment:
@jarry-xiao here we enforce that we only append a tree of the correct size given the nodes inserted into the tree so far

Reply:
Ah this is quite clever! Makes sense, because the intersection nodes maps to the root of the subtree where the newly appended node is the left most child